File: frontend/src/types/derived.ts
/**
 * Response envelope for derived (cached/computed) backend endpoints.
 *
 * Mirrors `backend/app/schemas/derived.py`. Applied today to
 * `/graph/stats` and `/graph/metadata/schema` — other derived endpoints
 * still return their raw shapes.
 *
 * The envelope exists so a cold or degraded cache yields a normal 200
 * with an explicit `status` the UI can handle, instead of a 504 that
 * breaks the page.
 */

export type DerivedStatus =
  | 'fresh'        // within TTL
  | 'stale'        // past TTL, usable, recompute queued
  | 'computing'    // no cached data, recompute queued
  | 'partial'      // schema only: types present, counts missing
  | 'unavailable'  // cache dead and fallback dead (rare)

export interface DerivedMeta {
  status: DerivedStatus
  source: 'memory' | 'db' | 'ontology' | 'none'
  /** ISO timestamp of last successful computation, if any. */
  computed_at: string | null
  age_seconds: number
  ttl_seconds: number
  job_id: string | null
  missing_fields: string[]
}

export interface DerivedResponse<T> {
  data: T | null
  meta: DerivedMeta
}

/**
 * Type guard: `true` when the value looks like a DerivedResponse envelope.
 *
 * Used by the fetch adapters to detect whether a given endpoint has been
 * migrated to the envelope shape, so callers can still consume legacy
 * raw responses without breaking.
 */
export function isDerivedResponse<T>(value: unknown): value is DerivedResponse<T> {
  if (!value || typeof value !== 'object') return false
  const v = value as { meta?: unknown; data?: unknown }
  if (v.meta === undefined || !v.meta || typeof v.meta !== 'object') return false
  const m = v.meta as { status?: unknown }
  return typeof m.status === 'string'
}

/**
 * Unwrap a DerivedResponse to its `data` payload, or pass through a
 * legacy raw response unchanged. Throws only on `status=unavailable`
 * (a genuine infra failure) so callers preserve their existing
 * error-handling semantics.
 */
export function unwrapDerived<T>(value: T | DerivedResponse<T>): {
  data: T | null
  meta: DerivedMeta | null
} {
  if (isDerivedResponse<T>(value)) {
    if (value.meta.status === 'unavailable') {
      throw new Error(
        `Derived endpoint returned status=unavailable (source=${value.meta.source}). ` +
        `Cache and fallback are both unreachable.`,
      )
    }
    return { data: value.data, meta: value.meta }
  }
  // Legacy raw response — no envelope, return as-is with null meta.
  return { data: value as T, meta: null }
}

/**
 * True iff the envelope represents a state where data is present but
 * might be behind the live graph (stale, partial, computing-with-prior-data).
 * Useful for UI banners that say "schema is being refreshed".
 */
export function isDerivedStale(meta: DerivedMeta | null): boolean {
  if (!meta) return false
  return meta.status === 'stale' || meta.status === 'partial' || meta.status === 'computing'
}

File: backend/app/schemas/derived.py
"""Response envelope for derived (cached/computed) endpoints.

Applied to `/graph/stats` and `/graph/metadata/schema` so a cold or
degraded cache never produces a 504 — every response is a normal 200
with an explicit ``status`` field that the frontend can handle.

Envelope is intentionally generic so other derived endpoints can adopt
it later with ``response_model=DerivedResponse[MyType]`` and no new
design work; today only the two problem endpoints use it.
"""
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum
from typing import Generic, List, Literal, Optional, TypeVar

from pydantic import BaseModel, Field


class DerivedStatus(str, Enum):
    FRESH = "fresh"               # within TTL
    STALE = "stale"               # past TTL, usable, recompute queued
    COMPUTING = "computing"       # no cached data, recompute queued
    PARTIAL = "partial"           # schema only: types present, counts missing
    UNAVAILABLE = "unavailable"   # cache dead and fallback dead (rare)


class DerivedMeta(BaseModel):
    status: DerivedStatus
    # Intentionally excludes "live": GET handlers must not call the provider.
    source: Literal["memory", "db", "ontology", "none"] = "none"
    computed_at: Optional[datetime] = None
    age_seconds: int = 0
    ttl_seconds: int = 0
    job_id: Optional[str] = None
    missing_fields: List[str] = Field(default_factory=list)


T = TypeVar("T")


class DerivedResponse(BaseModel, Generic[T]):
    data: Optional[T] = None
    meta: DerivedMeta


def _now() -> datetime:
    return datetime.now(timezone.utc)


def compute_age_seconds(updated_at_iso: Optional[str]) -> int:
    """Seconds since ``updated_at_iso`` (ISO string, timezone-aware or naive-UTC).

    Returns 0 if the timestamp is missing or unparseable — callers should
    treat that as "unknown age", not "fresh".
    """
    if not updated_at_iso:
        return 0
    try:
        ts = datetime.fromisoformat(updated_at_iso)
        if ts.tzinfo is None:
            ts = ts.replace(tzinfo=timezone.utc)
        return max(0, int((_now() - ts).total_seconds()))
    except (ValueError, TypeError):
        return 0


def classify_freshness(age_seconds: int, ttl_seconds: int) -> DerivedStatus:
    """``fresh`` if age ≤ TTL, ``stale`` otherwise. Caller picks TTL per endpoint."""
    return DerivedStatus.FRESH if age_seconds <= ttl_seconds else DerivedStatus.STALE
